### ISSUE TITLE
InfluxDB: High CPU utilization in v.1.8.1 & v.1.8.2

### ISSUE THREAD

**Issue by teemuku** [influxdata/influxdb#19543]
Experienced some performance issues with InfluxDB, after upgrading from InfluxDB v.1.8.0 to v1.8.1. As of v.1.8.2 has now been released, it has also been tested and the problem still exists. Currently the problem is temporarily handled by downgrading back to 1.8.0.

__Steps to reproduce:__
In this environment, nothing else than
1. Upgrade InfluxDB version to 1.8.1 or 1.8.2

__Expected behavior:__
InfluxDB updates to a newer version and continues to work without issues.

__Actual behavior:__
CPU Utilization spiking to ~100%, some databases are being queried successfully, where as at least one of the largest databases (~65GB) is returning HTTP POST 500 / timeout.
Other metrics, such as memory & disk metrics are not affected drastically.

Screenshots from Grafana to visualize the CPU utilization after upgrading to v.1.8.1
![v1 8 1_cpu_util](https://user-images.githubusercontent.com/70326653/92712755-c8d57b80-f362-11ea-89e8-76220dc73e01.JPG)

And after upgrading to v.1.8.2
![v1 8 2_cpu_util](https://user-images.githubusercontent.com/70326653/92712834-e276c300-f362-11ea-8587-fcdd16816271.JPG)

__Environment info:__

* System info: Linux 3.10.0-1127.18.2.el7.x86_64 x86_64
* InfluxDB version: InfluxDB v1.8.0 (git: 1.8 781490d)
* Host VMs specs:
CentOS 7
8 vCPUs
52 GB RAM
500 GB SSD disk
* Other environment details:
Two InfluxDB instances running on separate Linux VMs. No other special/heavy workloads running on the servers other than InfluxDB. Instances are not aware of each other, and they are identical in terms of configuration and data ingest. They are also being queried by the same Grafana frontend servers.
* InfluxDB instances:
~65 databases
~150 GB of data (~85% raw data, ~15% downsampled)
Config settings on default other than some directory settings and TSI indexing turned on
Wal and Data directories are located on the same storage device

__Logs:__
Example error line from journalctl:
```
 Sep 10 10:01:04 influxdb02 influxd[22667]: ts=2020-09-10T07:01:04.768040Z lvl=error msg="[500] - \"timeout\"" log_id=0P9h7VtW000 service=httpd
```

Example error line from HTTP access log:
```
x.x.x.x - telegraf [10/Sep/2020:10:01:00 +0300] "POST /write?db=all_operated HTTP/1.1" 500 20 "-" "Go-http-client/1.1" 62720985-f333-11ea-ac27-42010ae8030c 10609871
```

__Other notes:__
Not sure if this is relevant regarding to this issue, but there has also been some memory issues most likely caused by CQs, since disabling CQs for testing purposes resulted in stable memory usage. Each db has identical CQ and there has been spiking in memory usage, sometimes OoMs even, with version 1.8.0 too.
Retention policies in use:
```
> show retention policies
name    duration  shardGroupDuration replicaN default
----    --------  ------------------ -------- -------
autogen 0s        168h0m0s           1        false
raw     336h0m0s  24h0m0s            1        true
agg     9600h0m0s 168h0m0s           1        false
```
Data is downsampled from "raw" to "agg" RP with continuous queries.
```
> show continuous queries
name: <database>
name         query
----         -----
cq_aggregate CREATE CONTINUOUS QUERY cq_aggregate ON <database> BEGIN SELECT mean(*) INTO <database>.agg.:MEASUREMENT FROM <database>.raw./.*/ GROUP BY time(5m), * END
```

**Comment by teemuku** [comment]
I updated/added some information on this issue.

**Comment by Paraphraser** [comment]
I think I figured this one out. I'm running a MING stack under Docker on a 4GB Raspberry Pi 4 (via [SensorsIot/IOTstack](https://github.com/SensorsIot/IOTstack)) and had CPU utilisation from InfluxDB routinely sitting around 85%.

> This is Influx 1.8.3 (no Docker container for Influx 2.0 has come my way yet).

I have two identical RPis, one "live", one "test". The live RPi was ingesting data from MQTT via Node-Red at the rate of 600 insertions per hour. No continuous queries of my own. No retention policies of my own other than "keep everything". A few Grafana dashboards but even when none of those were open, I was still seeing numbers like 85%.

This was on **both** RPis, even though the "test" RPi wasn't ingesting any data nor responding to Grafana queries.

It made no sense.

It turns out to be the "_internal" database. For some reason, this takes a HUGE number of resources. It did seem to have a 24H retention policy so maybe it was all that gardening.

I added this environment key:

```
INFLUXDB_MONITOR_STORE_ENABLED=FALSE
```

then recreated the InfluxDB container and, bingo, CPU on the "live" RPi is in the 1..5% range unless I do something like ask Grafana to show me a graph of something "for the last year". That'll get a quick spike to 250% of CPU for InfluxDB but it will drop back to negligible as soon as I close the browser window.

CPU on the "test" RPi has numbers like 0.3%.

Also turns out that it is safe to `drop database _internal`. If you ever need to re-enable the setting, the `_internal` database gets auto-recreated.

I'll leave it up to People Who Know to figure out why the internal monitoring database is such a resource hog.

**Comment by Pabl0z** [comment]
I had similar issue with InfluxDB v1.8.3 (git: 1.8 563e6c3d1a7a2790763c6289501095dbec19244e)
on Raspberry Pi 3 Model B Plus Rev 1.3 - disabling monitoring and dropping _internal database helped [thanks @Paraphraser 👍]

`>show diagnostics
name: build
Branch Build Time Commit                                   Version
------ ---------- ------                                   -------
1.8               563e6c3d1a7a2790763c6289501095dbec19244e 1.8.3

name: config
bind-address   reporting-disabled
------------   ------------------
127.0.0.1:8088 false

name: config-coordinator
log-queries-after max-concurrent-queries max-select-buckets max-select-point max-select-series query-timeout write-timeout
----------------- ---------------------- ------------------ ---------------- ----------------- ------------- -------------
0s                0                      0                  0                0                 0s            10s

name: config-cqs
enabled query-stats-enabled run-interval
------- ------------------- ------------
true    false               1s

name: config-data
cache-max-memory-size cache-snapshot-memory-size cache-snapshot-write-cold-duration compact-full-write-cold-duration dir                    max-concurrent-compactions max-index-log-file-size max-series-per-database max-values-per-tag series-file-max-concurrent-compactions series-id-set-cache-size wal-dir               wal-fsync-delay
--------------------- -------------------------- ---------------------------------- -------------------------------- ---                    -------------------------- ----------------------- ----------------------- ------------------ -------------------------------------- ------------------------ -------               ---------------
1073741824            26214400                   10m0s                              4h0m0s                           /var/lib/influxdb/data 0                          1048576                 1000000                 100000             0                                      100                      /var/lib/influxdb/wal 0s

name: config-httpd
access-log-path bind-address enabled https-enabled max-connection-limit max-row-limit
--------------- ------------ ------- ------------- -------------------- -------------
                0.0.0.0:8086 true    false         0                    0

name: config-meta
dir
---
/var/lib/influxdb/meta

name: config-monitor
store-enabled
-------------
false

name: config-precreator
advance-period check-interval enabled
-------------- -------------- -------
30m0s          10m0s          true

name: config-retention
check-interval enabled
-------------- -------
30m0s          true

name: config-subscriber
enabled http-timeout write-buffer-size write-concurrency
------- ------------ ----------------- -----------------
true    30s          1000              40

name: network
hostname
--------
oh.local

name: runtime
GOARCH GOMAXPROCS GOOS  version
------ ---------- ----  -------
arm    4          linux go1.13.8

name: system
PID   currentTime                    started                        uptime
---   -----------                    -------                        ------
20780 2021-01-18T16:08:16.122554888Z 2021-01-18T16:03:26.002975412Z 4m50.119579476s
`

**Comment by anmenaga** [comment]
Had same problem on 4GB Pi4.
[This recommendation from documentation](https://docs.influxdata.com/platform/monitoring/influxdata-platform/internal-vs-external/#disable-the-internal-database-in-production-clusters) helped:
`To disable the _internal database, set "store-enabled" to "false" under the "[monitor]" section of your influxdb.conf.`

**Comment by Paraphraser** [comment]
Yes. To make sure anyone else reading this thread understands what is going on, the [two approaches](https://docs.influxdata.com/influxdb/v1.8/administration/config/#monitoring-settings) have exactly the same effect:

* *The configuration file approach:*
	* `influxdb.conf` → `[monitor]` → `store-enabled = false`

	If you're running InfluxDB as a "native" install where you can easily get to `influxdb.conf` then this approach is appropriate.

* *The [environment variable](https://docs.influxdata.com/influxdb/v1.8/administration/config/#influxdb-environment-variables-influxdb) approach:*
	* `INFLUXDB_MONITOR_STORE_ENABLED=FALSE`

	If you're running InfluxDB as a Docker container, it's a little trickier to get to the `influxdb.conf` file so this approach is more useful. There's a 1:1 relationship between config file settings and environment variables so you can always do what you want.

In any contest between the two approaches, the environment variable prevails.

**Comment by Waynedejager1** [comment]
The thread from here https://community.influxdata.com/t/strange-disk-space-usage-leak-after-influxdb-1-8-x-upgrade/15825/5
seems to pass blame on the entire v1.8x versions. 

I have set the monitor - store enabled to false and this definitely worked - albeit for 1 week. Now back to excessive CPU usage from Disk I/O reads....
Any other help guys?

**Comment by Waynedejager1** [comment]
By the way - i am using 1.8.0 and the same is happening

**Comment by Paraphraser** [comment]
I'm running 1.8.4 (although I'm not quite sure what that means - my docker-compose.yml pins to 1.8.4 but "pulls" still bring down updates from time to time so I think it might be a case of 1.8.4a, 1.8.4b, etc).

I have not seen any problem since I set monitor store enabled to false. Influx sits on a steady 1% of CPU, day in, day out. The Pi4 this is running on is rarely rebooted (typically weeks) so, if Influx was going to start going nuts it has plenty of opportunity. It isn't.

Don't know what else to suggest.

**Comment by Waynedejager1** [comment]
@Paraphraser , I rechecked the config file last night, turns out I forgot to uncomment the # ...duh. Running cool now and hoping it's fixed. 

Thank You ;-)

**Comment by Tatya-winchu** [comment]
I tried all the mentioned ways, i.e. `To disable the _internal database, set "store-enabled" to "false" under the "[monitor]" section of your influxdb.conf` and also downgrading the Influxdb to 1.7.x but no luck. 

After reading about Influxdb debug and CPU profiling HTTP API [HTTP API](https://www.alibabacloud.com/help/doc-detail/113115.htm) I was able to pin-down the issue, the problem was in the way I was making the query, my query involved more complex functions and also GROUP BAY tag.I also tried query analysis using ` EXPLAIN ANALYZE (query)` command to check how much time a query is taking to execute. I resolved that and noticed a huge Improvement in CPU load. Basically I can suggest the following:

    

1. Run the CPU profile analysis using influxdb HTTP API with the command `curl -o <file name> http://localhost:8086/debug/pprof/all?cpu=true` e and collect result.
2.  Visualize the result using go tool like [PPROFtool](https://github.com/google/pprof?spm=a2c63.p38356.879954.5.7aff6056WkxYmC) and find the problem
3. Identify the complex or moderate query using `EXPLAIN ANALYZE<query>` and try to improve the query format
4. Also get the information about number of request per second, series cardinality -> `SHOW SERIES CARDINALITY` 
5.  Before designing any schema and Influx client check the hardware recommendation ->> [Harware Guidelines](https://docs.influxdata.com/influxdb/v1.8/guides/hardware_sizing/#query-guidelines)

**Comment by jos149** [comment]
>     * _The [environment variable](https://docs.influxdata.com/influxdb/v1.8/administration/config/#influxdb-environment-variables-influxdb) approach:_
>       
>       * `INFLUXDB_MONITOR_STORE_ENABLED=FALSE`
>       
>       If you're running InfluxDB as a Docker container, it's a little trickier to get to the `influxdb.conf` file so this approach is more useful. There's a 1:1 relationship between config file settings and environment variables so you can always do what you want.
> 

Just to describe my approach of fixing this. I open portainer to enter the influx console. Once in the container console I do the following:
<img width="1384" alt="image" src="https://user-images.githubusercontent.com/41305927/135747904-903c04c7-b8a2-4d60-8d58-ef8ccc613e50.png">
What is wrong?

**Comment by Paraphraser** [comment]
You haven't given *quite* enough information for a definitive answer but, because you mention Portainer, I will assume you are running Influx as a Docker container **and** that you have a `docker-compose.yml` somewhere. See the following service definition taken from my own `docker-compose.yml` file:

```
  influxdb:
    container_name: influxdb
    image: "influxdb:1.8"
    restart: unless-stopped
    environment:
      - TZ=Australia/Sydney
      - INFLUXDB_HTTP_AUTH_ENABLED=false
      - INFLUXDB_HTTP_FLUX_ENABLED=false
      - INFLUXDB_REPORTING_DISABLED=false
      - INFLUXDB_MONITOR_STORE_ENABLED=FALSE
    ports:
      - "8086:8086"
    volumes:
      - ./volumes/influxdb/data:/var/lib/influxdb
      - ./backups/influxdb/db:/var/lib/influxdb/backup
```

Notice the `INFLUXDB_MONITOR_STORE_ENABLED` at the end of the `environment:` directive. That's where it goes - so that the variable is in effect as the container comes up. It can also go into an environment file. If you are using those, your service definition would have lines like this:

```
    env_file:
      - ./path/to/influxdb.env
```

and then the `influxdb.env` file would have a line that looked like this:

```
INFLUXDB_MONITOR_STORE_ENABLED=FALSE
```

I want to emphasise that I'm **not** telling you to use this exact service definition. I'm just trying to show you where it goes.

I **think** you can probably add environment variables in Portainer but I don't use it so I can't tell you how to do that.

The `INFLUXDB_MONITOR_STORE_ENABLED` is a directive to the `influxd` engine. That's not quite the same as `influx` which is the administrative CLI. And that's why it didn't have the effect you were looking for when you tried to do it via the console.

I hope this helps.

**Comment by jos149** [comment]
> You haven't given _quite_ enough information for a definitive answer but, because you mention Portainer, I will assume you are running Influx as a Docker container **and** that you have a `docker-compose.yml` somewhere. See the following service definition taken from my own `docker-compose.yml` file:
> 
> ```
>   influxdb:
>     container_name: influxdb
>     image: "influxdb:1.8"
>     restart: unless-stopped
>     environment:
>       - TZ=Australia/Sydney
>       - INFLUXDB_HTTP_AUTH_ENABLED=false
>       - INFLUXDB_HTTP_FLUX_ENABLED=false
>       - INFLUXDB_REPORTING_DISABLED=false
>       - INFLUXDB_MONITOR_STORE_ENABLED=FALSE
>     ports:
>       - "8086:8086"
>     volumes:
>       - ./volumes/influxdb/data:/var/lib/influxdb
>       - ./backups/influxdb/db:/var/lib/influxdb/backup
> ```
> 
> Notice the `INFLUXDB_MONITOR_STORE_ENABLED` at the end of the `environment:` directive. That's where it goes - so that the variable is in effect as the container comes up. It can also go into an environment file. If you are using those, your service definition would have lines like this:
> 
> ```
>     env_file:
>       - ./path/to/influxdb.env
> ```
> 
> and then the `influxdb.env` file would have a line that looked like this:
> 
> ```
> INFLUXDB_MONITOR_STORE_ENABLED=FALSE
> ```
> 
> I want to emphasise that I'm **not** telling you to use this exact service definition. I'm just trying to show you where it goes.
> 
> I **think** you can probably add environment variables in Portainer but I don't use it so I can't tell you how to do that.
> 
> The `INFLUXDB_MONITOR_STORE_ENABLED` is a directive to the `influxd` engine. That's not quite the same as `influx` which is the administrative CLI. And that's why it didn't have the effect you were looking for when you tried to do it via the console.
> 
> I hope this helps.

Thanks, I set it in the docker-compose.yml. Let's hope now my cpu usage goes down....

**Comment by jos149** [comment]
So my cpu usage is still very high. Never had this problem before although I did use the same config. What would be the next thing to do? I use a telegraf database in influxdb to plot data in Grafana. That's it.

**Comment by Paraphraser** [comment]
Did you also go into the CLI (as you did before via console) and run the command `drop database _internal`? It was the combination of the environment variable and dropping the `_internal` database that did the trick for me.

Also, I'm doing all this on Raspberry Pis. Very early on (~2019) I noticed a weird thing where queries on my "live" machine were increasingly sluggish while my "test" machine was instantaneous. I would routinely make a portable backup on the live machine and load it onto the test machine so, other than the live machine actually ingesting data from IoT devices (MQTT to NodeRED to Influx) there was no real difference in database size. One day, on a hunch, I wrote a script to run on the live machine: take a backup, erase everything, restore the backup. After that, the performance on the live machine came good and it has never shown any signs of degrading since then. I've always put it down to some glitch introduced by a previous version of the container, that had been cured by the time I wrote the script, and the full backup/restore removed whatever the glitch had introduced. But I don't know any of that as hard fact so maybe something similar would help you too.

If you want to give it a try, the script is at [this link](https://github.com/Paraphraser/IOTstackBackup/blob/master/scripts/iotstack_reload_influxdb) - but (a) it assumes the `docker-compose.yml` norms for [SensorsIot/IOTstack](https://github.com/SensorsIot/IOTstack), so (b) you should only treat it as a **source of ideas** and not try to run it as-is until **you** are 100% certain it is appropriate for your exact environment, and (c) it's entirely at your own risk (ie please don't blame me if it isn't suitable for your environment but you run it anyway and it winds up damaging your database).

**Comment by jos149** [comment]
> Did you also go into the CLI (as you did before via console) and run the command `drop database _internal`? It was the combination of the environment variable and dropping the `_internal` database that did the trick for me.

So just to summarize:
Go into the machine. Run `influx` and then just `drop database _internal`.
As far as I see this did not do anything....Sadly

**Comment by Paraphraser** [comment]
It's possible that it did actually "do nothing". It's also possible that it "did something" but you didn't realise it. As far as I can tell, the `_internal` database only gets created under the following conditions:

* `INFLUXDB_MONITOR_STORE_ENABLED` is either omitted or `TRUE`; **and**
* InfluxDB is initialised (in Docker that means starting from an empty persistent storage area).

If those conditions apply then you would expect to see:

```
$ docker exec -it influxdb bash
# echo $INFLUXDB_MONITOR_STORE_ENABLED
# influx
Connected to http://localhost:8086 version 1.8.9
InfluxDB shell version: 1.8.9
> show databases
name: databases
name
----
_internal
```

The `_internal` database doesn't go away if you change the environment variable:

```
$ docker exec -it influxdb bash
# echo $INFLUXDB_MONITOR_STORE_ENABLED
FALSE
# influx
Connected to http://localhost:8086 version 1.8.9
InfluxDB shell version: 1.8.9
> show databases
name: databases
name
----
_internal
```

But it does go away if you explicitly drop it:

```
> drop database _internal
> show databases
name: databases
name
----
```

Once `_internal` has been auto-created and then dropped, it doesn't seem to come back again even if you set `INFLUXDB_MONITOR_STORE_ENABLED` to `TRUE`. If you ever decide you want it back again, you can create it by hand via:

```
create database _internal
```

In some ways, Influx is a bit unusual. If you try commands like this:

```
create database fred
drop database fred
drop database fred
drop database jack
```

you don't get any response from any of them. If you intersperse those with `show databases` you can see `fred` after it is created, and then confirm it has disappeared, but dropping either an already-dropped database or a database that never existed doesn't produce any error messages. All the other database packages I've used would complain if I tried to drop something (database, table, etc) that didn't exist.

In other words, getting silence in response to `drop database _internal` doesn't tell you whether `_internal` did or did not exist prior to that point. 

Anyway, if you have tried all these tricks and you still have high CPU utilisation, the next thing I'd consider would be whether you had some hyper-active retention policies. That's something I can't really help with. I've taken the extremely lazy approach of not using any. My databases just keep growing. They've been accumulating data since 2015 and are currently at 536MB on an SSD which is only 3% full so it's going to be a while before I need to care too much.

**Comment by jos149** [comment]
Thanks for your explanation. That's a very nice overview. Easy to find out for me now. I will get back at this.

**Comment by jos149** [comment]
Took me a few days, however, still having the same issue. It appears nothing does the trick. The strange thing is that it never did this before. Now I am wondering whether it is possible to recreate the container. However, I suspect then my existing databases are gone :(.

**Comment by jos149** [comment]
Well, sadly enough this does not work either for me. But what I am seeing is that @Paraphraser is not running influx in a docker, correct?

**Comment by Paraphraser** [comment]
![8E91447C-AA07-4299-9733-AC81BC7D6899](https://user-images.githubusercontent.com/34226495/140648499-c053df8d-26b1-450b-aff7-05cc96f83e1e.jpeg)

**Comment by SergioRami** [comment]
Hi, 

Im currently running influxdb:1.8.10 on my Raspberry PI and did what @Paraphraser says about drop _internal db. But i still have CPU Usage above 100%. It happens during a TSM compacting file:

```
ts=2022-01-14T11:01:22.077655Z lvl=info msg="Error replacing new TSM files" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23oHEl000 op_name=tsm1_compact_group db_shard_id=262 error="cannot allocate memory"
ts=2022-01-14T11:01:23.078041Z lvl=info msg="TSM compaction (end)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23oHEl000 op_name=tsm1_compact_group db_shard_id=262 op_event=end op_elapsed=24266.339ms
ts=2022-01-14T11:01:23.811740Z lvl=info msg="TSM compaction (start)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 op_event=start
ts=2022-01-14T11:01:23.811799Z lvl=info msg="Beginning compaction" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_files_n=4
ts=2022-01-14T11:01:23.811823Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=0 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000009-000000002.tsm
ts=2022-01-14T11:01:23.811847Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=1 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000017-000000002.tsm
ts=2022-01-14T11:01:23.811870Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=2 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000025-000000002.tsm
ts=2022-01-14T11:01:23.811891Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=3 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000033-000000002.tsm
[httpd] 172.29.0.1 - telegraf [14/Jan/2022:11:01:24 +0000] "POST /write?db=home_assistant HTTP/1.1 " 204 0 "-" "python-requests/2.26.0" 5032d7cb-7529-11ec-803e-0242ac1d0006 86073
[httpd] 172.29.0.1 - telegraf [14/Jan/2022:11:01:32 +0000] "POST /write?consistency=any&db=db0 HTTP/1.1 " 204 0 "-" "Telegraf/1.20.2 Go/1.17" 55542ef8-7529-11ec-803f-0242ac1d0006 142324
[httpd] 172.29.0.1 - telegraf [14/Jan/2022:11:01:37 +0000] "POST /write?consistency=any&db=db0 HTTP/1.1 " 204 0 "-" "Telegraf/1.20.2 Go/1.17" 584e37ba-7529-11ec-8040-0242ac1d0006 206640
ts=2022-01-14T11:01:46.290035Z lvl=info msg="Error replacing new TSM files" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 error="cannot allocate memory"
ts=2022-01-14T11:01:47.290408Z lvl=info msg="TSM compaction (end)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 op_event=end op_elapsed=23478.667ms
ts=2022-01-14T11:01:47.811933Z lvl=info msg="TSM compaction (start)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 op_event=start
ts=2022-01-14T11:01:47.812020Z lvl=info msg="Beginning compaction" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_files_n=4
ts=2022-01-14T11:01:47.812060Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=0 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000009-000000002.tsm
ts=2022-01-14T11:01:47.812119Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=1 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000017-000000002.tsm
```

Anyone knows why is happening the error "Cannot allocate memory"?

**Comment by lanjia1028** [comment]
> Hi,
> 
> Im currently running influxdb:1.8.10 on my Raspberry PI 4 and did what @Paraphraser says about drop _internal db. But i still have CPU Usage above 100%. It happens during a TSM compacting file:
> 
> ```
> ts=2022-01-14T11:01:22.077655Z lvl=info msg="Error replacing new TSM files" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23oHEl000 op_name=tsm1_compact_group db_shard_id=262 error="cannot allocate memory"
> ts=2022-01-14T11:01:23.078041Z lvl=info msg="TSM compaction (end)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23oHEl000 op_name=tsm1_compact_group db_shard_id=262 op_event=end op_elapsed=24266.339ms
> ts=2022-01-14T11:01:23.811740Z lvl=info msg="TSM compaction (start)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 op_event=start
> ts=2022-01-14T11:01:23.811799Z lvl=info msg="Beginning compaction" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_files_n=4
> ts=2022-01-14T11:01:23.811823Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=0 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000009-000000002.tsm
> ts=2022-01-14T11:01:23.811847Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=1 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000017-000000002.tsm
> ts=2022-01-14T11:01:23.811870Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=2 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000025-000000002.tsm
> ts=2022-01-14T11:01:23.811891Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=3 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000033-000000002.tsm
> [httpd] 172.29.0.1 - telegraf [14/Jan/2022:11:01:24 +0000] "POST /write?db=home_assistant HTTP/1.1 " 204 0 "-" "python-requests/2.26.0" 5032d7cb-7529-11ec-803e-0242ac1d0006 86073
> [httpd] 172.29.0.1 - telegraf [14/Jan/2022:11:01:32 +0000] "POST /write?consistency=any&db=db0 HTTP/1.1 " 204 0 "-" "Telegraf/1.20.2 Go/1.17" 55542ef8-7529-11ec-803f-0242ac1d0006 142324
> [httpd] 172.29.0.1 - telegraf [14/Jan/2022:11:01:37 +0000] "POST /write?consistency=any&db=db0 HTTP/1.1 " 204 0 "-" "Telegraf/1.20.2 Go/1.17" 584e37ba-7529-11ec-8040-0242ac1d0006 206640
> ts=2022-01-14T11:01:46.290035Z lvl=info msg="Error replacing new TSM files" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 error="cannot allocate memory"
> ts=2022-01-14T11:01:47.290408Z lvl=info msg="TSM compaction (end)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23pntl000 op_name=tsm1_compact_group db_shard_id=262 op_event=end op_elapsed=23478.667ms
> ts=2022-01-14T11:01:47.811933Z lvl=info msg="TSM compaction (start)" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 op_event=start
> ts=2022-01-14T11:01:47.812020Z lvl=info msg="Beginning compaction" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_files_n=4
> ts=2022-01-14T11:01:47.812060Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=0 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000009-000000002.tsm
> ts=2022-01-14T11:01:47.812119Z lvl=info msg="Compacting file" log_id=0Z23ZeXG000 engine=tsm1 tsm1_level=2 tsm1_strategy=level trace_id=0Z23rGdl000 op_name=tsm1_compact_group db_shard_id=262 tsm1_index=1 tsm1_file=/var/lib/influxdb/data/db0/autogen/262/000000017-000000002.tsm
> ```
> 
> Anyone knows why is happening the error "Cannot allocate memory"?
same error , resolved by change para of linux.
vi /etc/sysctl.conf 
修改下述变量值
vm.max_map_count=4194304

sysctl -p

**Comment by Paraphraser** [comment]
Maybe see if you get any ideas from [this gist](https://gist.github.com/Paraphraser/f055815a6f26cb4e66e58138fa3fe4ae).

The background to the gist is that it was written to explore the effects of the `INFLUXDB_DATA_INDEX_VERSION` environment variable, but it's really about how that affects RAM utilisation.

It's also about InfluxDB 1.8 RAM utilisation more generally. I suggest you skim-read the bits about that environment variable and focus on how I've noticed that RAM growth can be curtailed by restarting the container every 24 hours. True, a restart can interrupt data sources so you might lose data if you don't (or can't) pick a good time of day. Only you can decide if that's a price worth paying.

Also, if you're running on an SSD, you might consider allocating a bit more swap space. I've upped mine to 2GB as a personal standard:

```
$ free -m
               total        used        free      shared  buff/cache   available
Mem:            7811         594         253           5        6963        7190
Swap:           2047           1        2046
```

It's rare to see it invaded on my 8GB Pi 4 but a bit more common on my 4GB Pis.

Hope this helps.
